### PR TITLE
run e2e tests under bazel

### PR DESCRIPTION
**Pull request by jbedard**
See the fixup commits for what changed since the last PR and what was reverted.

### CODE REVIEW

**Comment by alan-agius4** [comment]
@jbedard is this ready for review? As it's still marked as a draft.

**Comment by jbedard** [comment]
@alan-agius4 yes I believe this was ready.. 🤔

File: .circleci/dynamic_config.yml [hunk 1]
```diff
@@ -446,34 +560,63 @@ workflows:
...
+          snapshots: true
+          pre-steps:
+            - when:
+                condition:
+                  and:
+                    - not:
+                        equal: [main, << pipeline.git.branch >>]
+                    - not: << pipeline.parameters.snapshot_changed >>
+                steps:
+                  # Don't run snapshot E2E's unless it's on the main branch or the snapshots file has been updated.
```
Comment (alan-agius4) [inline .circleci/dynamic_config.yml:610]: The comment seems to be in the wrong place.

File: packages/angular_devkit/build_angular/src/builders/karma/index.ts [hunk 2]
```diff
@@ -222,6 +222,12 @@ function getBuiltInKarmaConfig(
     logLevel: karma.constants.LOG_INFO,
     autoWatch: true,
     browsers: ['Chrome'],
+    customLaunchers: {
```
Comment (alan-agius4) [inline packages/angular_devkit/build_angular/src/builders/karma/index.ts:225]: Can you add a comment hear that this is needed for bazel tests?

File: tests/legacy-cli/e2e/tests/generate/confg/type-karma.ts [hunk 3]
```diff
@@ -1,6 +1,8 @@
 import { ng } from '../../../utils/process';
+import { useCIChrome } from '../../../utils/project';
 
 export default async function () {
   await ng('generate', 'config', 'karma');
+  await useCIChrome('test-project');
```
Comment (jbedard) [inline tests/legacy-cli/e2e/tests/generate/confg/type-karma.ts:6]: This test must be new? This is required to update the karma config to use the headless chrome. See `project.ts`

File: tests/legacy-cli/e2e/utils/project.ts [hunk 4]
```diff
@@ -184,40 +189,45 @@ export function useCIDefaults(projectName = 'test-project'): Promise<void> {
...
       'directConnect: true,',
-      `directConnect: true, chromeDriver: String.raw\`${chromeDriverPath}\`,`,
+      `directConnect: true, chromeDriver: String.raw\`${process.env.CHROMEDRIVER_BIN}\`,`,
+    );
+  }
+
+  const karmaConf = path.join(projectDir, 'karma.conf.js');
+  if (fs.existsSync(karmaConf)) {
+    // Ensure the headless sandboxed chrome is configured in the karma config
+    await replaceInFile(
```
Comment (jbedard) [inline tests/legacy-cli/e2e/utils/project.ts:213]: This was required for the `generate/confg/type-karma.ts` test that creates a new karma config.

**Comment by jbedard** [comment]
Updated, see the last two fixup commits (https://github.com/angular/angular-cli/pull/24338/commits/5923c93ece1803c76badf82f1acfec440974d6f7, https://github.com/angular/angular-cli/pull/24338/commits/0259fd96c9ef5442b8e9cbea430ca024fd754f5a).

**Review by alan-agius4 — APPROVED** [review]

**Comment by alan-agius4** [comment]
@josephperrott PTAL.

**Review by josephperrott — APPROVED** [review]
LGTM

**Pull request merged**